# Import libraries

In [24]:
#import libraries
import pandas as pd
import numpy as np
import os

# Turning project folder path into a string

In [25]:
path = r'/Users/jorisjan/Documents/CareerFoundry/Instacart Basket Analysis'

# Import files into Jupyter as df_prods

In [26]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [27]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [28]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

# Data Wrangling

In [29]:
# Droping eval_set column from orders.csv
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [30]:
df_ords_2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [31]:
# count values of days_since_prior_order column
df_ords_2['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [32]:
# renaming order_dow column
df_ords_2.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [33]:
# Change Data Type of order_id and user_id
df_ords_2['order_id'] = df_ords['order_id'].astype('str')
df_ords_2['user_id'] = df_ords['user_id'].astype('str')

In [34]:
df_ords_2.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

In [35]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [36]:
# Transposing data in df_dep
df_dep_t = df_dep.T

In [37]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [38]:
# Adding an index to df_dep_t
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [39]:
# Create a new header for df_dep_t

# Create a new header
new_header = df_dep_t.iloc[0]

# Remove the first row in the dataframe
df_dep_t_new = df_dep_t[1:]

# Add the new header
df_dep_t_new.columns = new_header

In [40]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [41]:
# Turn df_dep_t_new into a dictionary
data_dict = df_dep_t_new.to_dict('index')

In [42]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [43]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [44]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [47]:
# Subsetting snack in products in products.csv
df_snacks =  df_prods[df_prods['department_id']==19]

In [48]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

# 2. Find another identifier variable and change Data Type as String

In [49]:
# Change Data Type of user_id
df_ords_2['user_id'] = df_ords['user_id'].astype('str')

# 3. Change name of 'days_since_prior_order' without overwriting the dataframe

In [50]:
# renaming days_since_prior_order column
df_ords_3 = df_ords_2.rename(columns = {'days_since_prior_order' : 'days_since_last_order'})

In [51]:
df_ords_3

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_last_order  
0                          NaN  
1                         15.0  
2                         21.0  
3                         29.0  
4                         28.0  
...                        ...  
3421078                   29.0  
3421079                   30.0  
3421080                   18.0  
3421081                    7.0  
3421082                   30.0  

[3421083 rows x 6 columns]

# 4. What is the busiest hour for placing orders?

In [53]:
df_ords_3['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

The busiest hour for placing orders is 10 o'clock !

# 5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [54]:
print(data_dict.get('4'))

{'department': 'produce'}


The meaning behind a value of 4 is 'produce'

# 6. Create a subset containing only the breakfast informations.

In [55]:
# Subsetting breakfast in products.csv
df_breakfast =  df_prods[df_prods['department_id']==14]

In [56]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [57]:
df_breakfast.describe()

product_id     aisle_id  department_id       prices
count   1116.000000  1116.000000         1116.0  1116.000000
mean   25464.490143   101.517921           14.0     7.848208
std    14271.353221    33.890362            0.0     3.971678
min       28.000000    48.000000           14.0     1.000000
25%    12754.250000    57.000000           14.0     4.400000
50%    26268.000000   121.000000           14.0     7.700000
75%    38035.250000   130.000000           14.0    11.300000
max    49663.000000   130.000000           14.0    14.900000

# 7. Create a subset containing informations about: alcohol, deli, beverages, and meat/seafood

In [58]:
# Subsetting dinner_parties in products.csv
df_dinner_parties =  df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [59]:
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [60]:
df_dinner_parties.describe()

product_id     aisle_id  department_id       prices
count   7650.000000  7650.000000    7650.000000  7650.000000
mean   24721.196601    66.286536       9.563268     9.012458
std    14297.565684    36.844580       5.114123     4.997438
min        3.000000     1.000000       5.000000     1.000000
25%    12402.500000    28.000000       7.000000     5.000000
50%    24803.000000    77.000000       7.000000     8.800000
75%    36977.250000    98.000000      12.000000    12.400000
max    49684.000000   134.000000      20.000000    25.000000

# 8. How many rows does the last dataframe you created have?

In [61]:
df_dinner_parties.shape

(7650, 5)

It has 7650 rows

# 9. Extract all the information you can about the user with a "user_id" of “1.”

In [68]:
# Subsetting user1 in products.csv
df_user1 =  df_ords_2[df_ords_2['user_id'] == '1']

In [69]:
df_user1

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# 10. Basic statistics and insights about the user's behavior

In [70]:
df_user1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

some basic statistics and insights about the user's behavior include Order Frequency and Timing (Total Orders, Orders by Day of Week, Orders by Hour of Day), Days Since Prior Order (Average Days Between Orders, Minimum Days Between Orders, Maximum Days Between Orders), Order Size and Growth, Data Completeness, Consistency…

# 12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [71]:
df_ords_2.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

# 13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder

In [73]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))